File: source/daemon/crates/wctl/src/commands/backup.rs
//! `wctl backup status | export | import | snapshots`.

use std::io::Read;
use std::path::Path;

use wardnet_common::api::{
    ApplyImportRequest, ApplyImportResponse, BackupStatusResponse, ExportBackupRequest,
    ListSnapshotsResponse, RestorePreviewResponse,
};
use wardnet_common::backup::{BackupStatus, MIN_PASSPHRASE_LEN, RestorePhase};

use crate::client::Client;
use crate::error::CliError;
use crate::output;

/// `GET /api/backup/status`.
///
/// # Errors
/// Propagates any [`CliError`] from the request.
pub async fn status(client: &Client, json: bool) -> Result<(), CliError> {
    let resp: BackupStatusResponse = client.get("/api/backup/status").await?;
    if json {
        return output::print_json(&resp);
    }
    println!("{}", backup_state(&resp.status));
    Ok(())
}

/// `POST /api/backup/export` — write the encrypted bundle to `out`.
///
/// # Errors
/// [`CliError::Usage`] for a too-short passphrase, [`CliError::Io`] on a write
/// failure, or any [`CliError`] from the request.
pub async fn export(
    client: &Client,
    json: bool,
    out: &str,
    passphrase_file: Option<&str>,
) -> Result<(), CliError> {
    let passphrase = read_passphrase(passphrase_file, true)?;
    if passphrase.chars().count() < MIN_PASSPHRASE_LEN {
        return Err(CliError::Usage(format!(
            "passphrase must be at least {MIN_PASSPHRASE_LEN} characters"
        )));
    }

    let req = ExportBackupRequest { passphrase };
    let bytes = client.post_json_bytes("/api/backup/export", &req).await?;
    std::fs::write(out, &bytes)
        .map_err(|e| CliError::Io(format!("cannot write bundle to {out}: {e}")))?;

    let written = bytes.len() as u64;
    if json {
        return output::print_json(&serde_json::json!({ "out": out, "bytes": written }));
    }
    println!("wrote {} to {out}", output::bytes(written));
    Ok(())
}

/// `POST /api/backup/import/preview` then `/apply` — restore a bundle.
///
/// The preview step is what validates the passphrase and compatibility before
/// anything on disk changes; only a compatible preview is applied.
///
/// # Errors
/// [`CliError::Io`] if the bundle cannot be read, [`CliError::Protocol`] if the
/// bundle is incompatible, or any [`CliError`] from the requests.
pub async fn import(
    client: &Client,
    json: bool,
    bundle_path: &str,
    passphrase_file: Option<&str>,
) -> Result<(), CliError> {
    let passphrase = read_passphrase(passphrase_file, false)?;
    let data = std::fs::read(bundle_path)
        .map_err(|e| CliError::Io(format!("cannot read bundle {bundle_path}: {e}")))?;
    let file_name = Path::new(bundle_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("bundle.wardnet.age")
        .to_owned();

    let form = reqwest::multipart::Form::new()
        .part(
            "bundle",
            reqwest::multipart::Part::bytes(data).file_name(file_name),
        )
        .text("passphrase", passphrase);
    let preview: RestorePreviewResponse = client
        .post_multipart("/api/backup/import/preview", form)
        .await?;

    if !preview.compatible {
        let reason = preview
            .incompatibility_reason
            .unwrap_or_else(|| "incompatible bundle".to_owned());
        return Err(CliError::Protocol(format!(
            "bundle cannot be restored: {reason}"
        )));
    }

    if !json {
        println!(
            "restoring bundle from host {} (schema v{})",
            preview.manifest.host_id, preview.manifest.schema_version
        );
        for file in &preview.files_to_replace {
            println!("  will replace: {file}");
        }
    }

    let apply_req = ApplyImportRequest {
        preview_token: preview.preview_token,
    };
    let applied: ApplyImportResponse = client
        .post_json("/api/backup/import/apply", &apply_req)
        .await?;

    if json {
        return output::print_json(&applied);
    }
    println!("restore applied; restart the daemon to load the new state");
    for snap in &applied.snapshots {
        println!("  snapshot: {} ({})", snap.path, snap.kind.as_str());
    }
    Ok(())
}

/// `GET /api/backup/snapshots`.
///
/// # Errors
/// Propagates any [`CliError`] from the request.
pub async fn snapshots(client: &Client, json: bool) -> Result<(), CliError> {
    let resp: ListSnapshotsResponse = client.get("/api/backup/snapshots").await?;
    if json {
        return output::print_json(&resp);
    }
    if resp.snapshots.is_empty() {
        println!("no snapshots");
        return Ok(());
    }
    println!("{:<10}  {:<12}  {:<26}  PATH", "KIND", "SIZE", "CREATED");
    for snap in &resp.snapshots {
        println!(
            "{:<10}  {:<12}  {:<26}  {}",
            snap.kind.as_str(),
            output::bytes(snap.size_bytes),
            snap.created_at.to_rfc3339(),
            snap.path,
        );
    }
    Ok(())
}

/// Obtain the export/import passphrase.
///
/// From `--passphrase-file` (a path, or `-` for stdin) when given, otherwise a
/// no-echo interactive prompt. When `confirm` is set (export), the prompt is
/// asked twice and the two must match.
fn read_passphrase(passphrase_file: Option<&str>, confirm: bool) -> Result<String, CliError> {
    if let Some(source) = passphrase_file {
        let raw = if source == "-" {
            let mut buf = String::new();
            std::io::stdin()
                .read_to_string(&mut buf)
                .map_err(|e| CliError::Io(format!("reading passphrase from stdin failed: {e}")))?;
            buf
        } else {
            std::fs::read_to_string(source)
                .map_err(|e| CliError::Io(format!("cannot read passphrase file {source}: {e}")))?
        };
        // Strip only a trailing newline (and CR) that `echo`/editors append —
        // interior characters are part of the passphrase.
        return Ok(raw.trim_end_matches(['\r', '\n']).to_owned());
    }

    let pass = rpassword::prompt_password("Passphrase: ")
        .map_err(|e| CliError::Io(format!("reading passphrase failed: {e}")))?;
    if confirm {
        let again = rpassword::prompt_password("Confirm passphrase: ")
            .map_err(|e| CliError::Io(format!("reading passphrase failed: {e}")))?;
        if pass != again {
            return Err(CliError::Usage("passphrases do not match".to_owned()));
        }
    }
    Ok(pass)
}

fn backup_state(status: &BackupStatus) -> String {
    match status {
        BackupStatus::Idle => "idle".to_owned(),
        BackupStatus::Exporting => "exporting".to_owned(),
        BackupStatus::Importing { phase } => format!("importing ({})", restore_phase(phase)),
        BackupStatus::Failed { reason } => format!("failed: {reason}"),
    }
}

fn restore_phase(phase: &RestorePhase) -> String {
    match phase {
        RestorePhase::Idle => "idle".to_owned(),
        RestorePhase::Validating => "validating".to_owned(),
        RestorePhase::StoppingRunners => "stopping-runners".to_owned(),
        RestorePhase::BackingUp => "backing-up".to_owned(),
        RestorePhase::Extracting => "extracting".to_owned(),
        RestorePhase::Migrating => "migrating".to_owned(),
        RestorePhase::RestartingRunners => "restarting-runners".to_owned(),
        RestorePhase::Applied => "applied".to_owned(),
        RestorePhase::Failed { reason } => format!("failed: {reason}"),
    }
}

File: source/wctl/cmd/wctl/backup.go
package main

import (
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/spf13/cobra"
	"golang.org/x/term"
	wardnet "wardnet.network/go"
)

// minPassphraseLen mirrors the daemon's backup passphrase minimum so a
// too-short passphrase is rejected before any request.
const minPassphraseLen = 12

func newBackupCmd(c *cli) *cobra.Command {
	cmd := &cobra.Command{Use: "backup", Short: "Export and restore encrypted backup bundles"}
	cmd.AddCommand(
		backupStatusCmd(c),
		backupExportCmd(c),
		backupImportCmd(c),
		backupSnapshotsCmd(c),
	)
	return cmd
}

func backupStatusCmd(c *cli) *cobra.Command {
	return &cobra.Command{
		Use:   "status",
		Short: "Show the current backup subsystem phase",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			client, err := c.client()
			if err != nil {
				return err
			}
			s, err := client.Backup.Status(cmd.Context())
			if err != nil {
				return err
			}
			if c.jsonOut {
				return printJSON(s)
			}
			fmt.Println(backupStatusString(s))
			return nil
		},
	}
}

func backupStatusString(s *wardnet.BackupStatus) string {
	switch s.State {
	case wardnet.BackupImporting:
		return fmt.Sprintf("importing (%s)", s.Phase)
	case wardnet.BackupFailed:
		return "failed: " + s.Reason
	default:
		return string(s.State)
	}
}

func backupExportCmd(c *cli) *cobra.Command {
	var out, passphraseFile string
	cmd := &cobra.Command{
		Use:   "export",
		Short: "Export an encrypted bundle to --out",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			passphrase, err := readPassphrase(passphraseFile, true)
			if err != nil {
				return err
			}
			if len(passphrase) < minPassphraseLen {
				return fmt.Errorf("passphrase must be at least %d characters", minPassphraseLen)
			}
			client, err := c.client()
			if err != nil {
				return err
			}
			bundle, err := client.Backup.Export(cmd.Context(), passphrase)
			if err != nil {
				return err
			}
			if err := os.WriteFile(out, bundle, 0o600); err != nil {
				return fmt.Errorf("cannot write bundle to %s: %w", out, err)
			}
			if c.jsonOut {
				return printJSON(map[string]any{"out": out, "bytes": len(bundle)})
			}
			fmt.Printf("wrote %s to %s\n", humanBytes(int64(len(bundle))), out)
			return nil
		},
	}
	cmd.Flags().StringVar(&out, "out", "", "destination path for the bundle")
	cmd.Flags().StringVar(&passphraseFile, "passphrase-file", "",
		"read the passphrase from this file instead of prompting (use - for stdin)")
	_ = cmd.MarkFlagRequired("out")
	return cmd
}

func backupImportCmd(c *cli) *cobra.Command {
	var passphraseFile string
	cmd := &cobra.Command{
		Use:   "import <bundle>",
		Short: "Restore a previously-exported bundle",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			passphrase, err := readPassphrase(passphraseFile, false)
			if err != nil {
				return err
			}
			data, err := os.ReadFile(args[0])
			if err != nil {
				return fmt.Errorf("cannot read bundle %s: %w", args[0], err)
			}
			client, err := c.client()
			if err != nil {
				return err
			}
			preview, err := client.Backup.PreviewImport(cmd.Context(), data, passphrase)
			if err != nil {
				return err
			}
			if !preview.Compatible {
				reason := preview.IncompatibilityReason
				if reason == "" {
					reason = "incompatible bundle"
				}
				return fmt.Errorf("bundle cannot be restored: %s", reason)
			}
			if !c.jsonOut {
				fmt.Printf("restoring bundle from host %s (schema v%d)\n",
					preview.Manifest.HostID, preview.Manifest.SchemaVersion)
				for _, f := range preview.FilesToReplace {
					fmt.Printf("  will replace: %s\n", f)
				}
			}
			result, err := client.Backup.ApplyImport(cmd.Context(), preview.PreviewToken)
			if err != nil {
				return err
			}
			if c.jsonOut {
				return printJSON(result)
			}
			fmt.Println("restore applied; restart the daemon to load the new state")
			for _, s := range result.Snapshots {
				fmt.Printf("  snapshot: %s (%s)\n", s.Path, s.Kind)
			}
			return nil
		},
	}
	cmd.Flags().StringVar(&passphraseFile, "passphrase-file", "",
		"read the passphrase from this file instead of prompting (use - for stdin)")
	return cmd
}

func backupSnapshotsCmd(c *cli) *cobra.Command {
	return &cobra.Command{
		Use:   "snapshots",
		Short: "List retained restore snapshots",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			client, err := c.client()
			if err != nil {
				return err
			}
			snaps, err := client.Backup.Snapshots(cmd.Context())
			if err != nil {
				return err
			}
			if c.jsonOut {
				return printJSON(snaps)
			}
			if len(snaps) == 0 {
				fmt.Println("no snapshots")
				return nil
			}
			tw := newTabWriter()
			fmt.Fprintln(tw, "KIND\tSIZE\tCREATED\tPATH")
			for i := range snaps {
				s := &snaps[i]
				fmt.Fprintf(tw, "%s\t%s\t%s\t%s\n",
					s.Kind, humanBytes(s.SizeBytes),
					s.CreatedAt.Format("2006-01-02T15:04:05Z07:00"), s.Path)
			}
			return tw.Flush()
		},
	}
}

// readPassphrase obtains the export/import passphrase from a file (or "-" for
// stdin) when passphraseFile is set, otherwise via a no-echo prompt. When
// confirm is set (export), the prompt is asked twice and the two must match.
func readPassphrase(passphraseFile string, confirm bool) (string, error) {
	if passphraseFile != "" {
		var raw []byte
		var err error
		if passphraseFile == "-" {
			raw, err = io.ReadAll(os.Stdin)
		} else {
			raw, err = os.ReadFile(passphraseFile)
		}
		if err != nil {
			return "", fmt.Errorf("cannot read passphrase: %w", err)
		}
		return strings.TrimRight(string(raw), "\r\n"), nil
	}

	fmt.Fprint(os.Stderr, "Passphrase: ")
	first, err := term.ReadPassword(int(os.Stdin.Fd()))
	fmt.Fprintln(os.Stderr)
	if err != nil {
		return "", fmt.Errorf("reading passphrase: %w", err)
	}
	if confirm {
		fmt.Fprint(os.Stderr, "Confirm passphrase: ")
		again, err := term.ReadPassword(int(os.Stdin.Fd()))
		fmt.Fprintln(os.Stderr)
		if err != nil {
			return "", fmt.Errorf("reading passphrase: %w", err)
		}
		if string(first) != string(again) {
			return "", fmt.Errorf("passphrases do not match")
		}
	}
	return string(first), nil
}
